File: apps/ios-native/Sources/WensenScreen.swift
import SwiftUI

struct WensenScreen: View {
    @EnvironmentObject var session: Session

    @State private var requests: [FeatureRequest] = []
    @State private var selectedStatus: String? = nil   // nil = "Populair"
    @State private var loading = false
    @State private var errorMessage: String? = nil
    @State private var showNew = false
    @State private var showSignIn = false

    private struct StatusFilter: Identifiable {
        let id: String           // unique key used in ForEach
        let status: String?      // nil = no filter (popular)
        let label: String
        var isSelected: Bool
    }

    private var filters: [StatusFilter] {
        [
            StatusFilter(id: "all",         status: nil,           label: "Populair",      isSelected: selectedStatus == nil),
            StatusFilter(id: "CONSIDERING", status: "CONSIDERING", label: "In overweging", isSelected: selectedStatus == "CONSIDERING"),
            StatusFilter(id: "PLANNED",     status: "PLANNED",     label: "Gepland",       isSelected: selectedStatus == "PLANNED"),
            StatusFilter(id: "DONE",        status: "DONE",        label: "Klaar",         isSelected: selectedStatus == "DONE"),
            StatusFilter(id: "DECLINED",    status: "DECLINED",    label: "Afgewezen",     isSelected: selectedStatus == "DECLINED"),
        ]
    }

    var body: some View {
        NavigationStack {
            ZStack(alignment: .bottomTrailing) {
                Brand.sand.ignoresSafeArea()

                VStack(spacing: 0) {
                    filterRow
                    contentArea
                }

                fab
                    .padding([.bottom, .trailing], DVH.s5)
            }
            .navigationTitle("Wensen")
            .navigationBarTitleDisplayMode(.inline)
        }
        .sheet(isPresented: $showNew) {
            RequestNewView { newRequest in
                requests.insert(newRequest, at: 0)
            }
        }
        .sheet(isPresented: $showSignIn) {
            SignInView(
                reason: "Log in om een wens in te dienen of om te stemmen.",
                dismissable: true
            )
            .environmentObject(session)
        }
        .task { await loadRequests() }
    }

    // MARK: - Filter chips

    private var filterRow: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: DVH.s2) {
                ForEach(filters) { filter in
                    DVHChip(
                        label: filter.label,
                        selected: filter.isSelected
                    ) {
                        if selectedStatus != filter.status {
                            selectedStatus = filter.status
                            Task { await loadRequests() }
                        }
                    }
                }
            }
            .padding(.horizontal, DVH.s5)
            .padding(.vertical, DVH.s2)
        }
    }

    // MARK: - Content

    @ViewBuilder
    private var contentArea: some View {
        if loading && requests.isEmpty {
            Spacer()
            ProgressView()
                .frame(maxWidth: .infinity)
            Spacer()
        } else if let msg = errorMessage {
            EmptyStateView(
                icon: "exclamationmark.circle",
                title: "Kon wensen niet laden",
                message: msg,
                actionLabel: "Opnieuw proberen"
            ) {
                Task { await loadRequests() }
            }
        } else if requests.isEmpty {
            EmptyStateView(
                icon: "lightbulb",
                title: "Nog geen wensen",
                message: "Wees de eerste die een idee indient.",
                actionLabel: "Idee indienen"
            ) {
                handleFABTap()
            }
        } else {
            ScrollView {
                LazyVStack(spacing: DVH.s3) {
                    ForEach(requests) { req in
                        NavigationLink(destination: RequestDetailView(request: req)) {
                            WishCard(request: req)
                        }
                        .buttonStyle(.plain)
                    }
                }
                .padding(.horizontal, DVH.s5)
                .padding(.top, DVH.s3)
                .padding(.bottom, 100)   // clear FAB
            }
            .refreshable { await loadRequests() }
        }
    }

    // MARK: - FAB

    private var fab: some View {
        Button { handleFABTap() } label: {
            Image(systemName: "plus")
                .font(.title2.weight(.semibold))
                .foregroundStyle(.white)
                .frame(width: 56, height: 56)
                .background(Brand.moss, in: Circle())
                .shadow(color: Brand.moss.opacity(0.35), radius: 12, y: 4)
        }
    }

    // MARK: - Actions

    private func handleFABTap() {
        if session.isAuthenticated {
            showNew = true
        } else {
            showSignIn = true
        }
    }

    private func loadRequests() async {
        loading = true
        errorMessage = nil
        do {
            requests = try await APIClient.featureRequests(status: selectedStatus)
        } catch {
            errorMessage = "Er is iets misgegaan. Probeer het opnieuw."
        }
        loading = false
    }
}

// MARK: - WishCard

private struct WishCard: View {
    let request: FeatureRequest

    var body: some View {
        HStack(alignment: .top, spacing: DVH.s3) {
            VStack(alignment: .leading, spacing: DVH.s2) {
                Text(request.title)
                    .font(.dvhHeadline)
                    .foregroundStyle(Brand.ink)
                    .multilineTextAlignment(.leading)

                if let body = request.body, !body.isEmpty {
                    Text(body)
                        .font(.dvhBody)
                        .foregroundStyle(Brand.ink2)
                        .lineLimit(2)
                        .multilineTextAlignment(.leading)
                }

                if let component = request.component, !component.isEmpty {
                    WishComponentTag(label: component)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)

            upvoteColumn
        }
        .dvhCard()
    }

    private var upvoteColumn: some View {
        VStack(spacing: DVH.s1) {
            Image(systemName: "chevron.up.circle")
                .font(.system(size: 22, weight: .medium))
                .foregroundStyle(request.viewerHasVoted == true ? Brand.moss : Brand.ink2)
            Text("\(request.upvoteCount ?? 0)")
                .font(.dvhCaption.weight(.semibold))
                .foregroundStyle(Brand.ink2)
        }
    }
}

// MARK: - WishComponentTag

private struct WishComponentTag: View {
    let label: String

    var body: some View {
        Text(label)
            .font(.dvhCaption.weight(.medium))
            .foregroundStyle(Brand.mossDark)
            .padding(.horizontal, DVH.s2 + 2)
            .padding(.vertical, DVH.s1)
            .background(Brand.mossSoft.opacity(0.7), in: Capsule())
    }
}

File: apps/android/app/src/main/java/nl/devrijehond/app/ui/wensen/WensenScreen.kt
package nl.devrijehond.app.ui.wensen

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.KeyboardArrowUp
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.FloatingActionButton
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TextFieldDefaults
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.lifecycle.viewmodel.compose.viewModel
import kotlinx.coroutines.launch
import nl.devrijehond.app.AppGraph
import nl.devrijehond.app.api.models.FeatureRequest
import nl.devrijehond.app.api.models.FeatureStatus
import nl.devrijehond.app.ui.theme.Brand
import nl.devrijehond.app.ui.theme.Dvh

/**
 * Wensen tab: community feature requests. Anyone can browse and filter; voting and
 * submitting require sign-in (handled via [onRequireSignIn]). Mirrors the iOS
 * WensenScreen.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun WensenScreen(
    onRequireSignIn: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val vm: WensenViewModel = viewModel()
    val state by vm.state.collectAsState()
    val token by AppGraph.session.token.collectAsState()
    val isAuthenticated = token != null

    var showNew by remember { mutableStateOf(false) }
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    val scope = rememberCoroutineScope()

    Scaffold(
        modifier = modifier,
        containerColor = Brand.Sand,
        floatingActionButton = {
            FloatingActionButton(
                onClick = { if (isAuthenticated) showNew = true else onRequireSignIn() },
                containerColor = Brand.Moss,
                contentColor = Color.White,
            ) {
                Icon(Icons.Filled.Add, contentDescription = "Idee indienen")
            }
        },
    ) { innerPadding ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(innerPadding),
        ) {
            Text(
                text = "Wensen",
                style = MaterialTheme.typography.displaySmall,
                color = Brand.Ink,
                modifier = Modifier.padding(start = Dvh.s4, top = Dvh.s4),
            )
            Text(
                text = "Stem op ideeën of dien er zelf een in.",
                style = MaterialTheme.typography.bodyMedium,
                color = Brand.Ink2,
                modifier = Modifier.padding(start = Dvh.s4, top = Dvh.s1, bottom = Dvh.s3),
            )

            FilterRow(selected = state.filter, onSelect = vm::selectFilter)

            when {
                state.loading && state.requests.isEmpty() -> CenterBox {
                    CircularProgressIndicator(color = Brand.Moss)
                }

                state.error != null && state.requests.isEmpty() -> CenterBox {
                    Column(horizontalAlignment = Alignment.CenterHorizontally) {
                        Text(state.error!!, color = Brand.Rust)
                        Button(
                            onClick = vm::load,
                            modifier = Modifier.padding(top = Dvh.s3),
                            colors = ButtonDefaults.buttonColors(containerColor = Brand.Moss),
                        ) { Text("Opnieuw") }
                    }
                }

                state.requests.isEmpty() -> CenterBox {
                    Column(horizontalAlignment = Alignment.CenterHorizontally) {
                        Text("Nog geen wensen", color = Brand.Ink, fontWeight = FontWeight.SemiBold)
                        Text(
                            "Wees de eerste die een idee indient.",
                            color = Brand.Ink2,
                            modifier = Modifier.padding(top = Dvh.s1),
                        )
                    }
                }

                else -> LazyColumn(
                    contentPadding = androidx.compose.foundation.layout.PaddingValues(
                        start = Dvh.s4,
                        end = Dvh.s4,
                        top = Dvh.s3,
                        bottom = 96.dp,
                    ),
                    verticalArrangement = Arrangement.spacedBy(Dvh.s3),
                ) {
                    items(state.requests, key = { it.id.toString() }) { req ->
                        WishCard(
                            request = req,
                            voting = state.votingIds.contains(req.id.toString()),
                            onUpvote = {
                                if (isAuthenticated) vm.toggleVote(req) else onRequireSignIn()
                            },
                        )
                    }
                }
            }
        }
    }

    if (showNew) {
        ModalBottomSheet(
            onDismissRequest = { showNew = false },
            sheetState = sheetState,
            containerColor = Brand.Cream,
        ) {
            NewWishSheet(
                onSubmit = { title, body, component, onResult ->
                    vm.create(title, body, component) { error ->
                        if (error == null) {
                            scope.launch {
                                sheetState.hide()
                                showNew = false
                            }
                        }
                        onResult(error)
                    }
                },
            )
        }
    }
}

@Composable
private fun FilterRow(
    selected: FeatureStatus?,
    onSelect: (FeatureStatus?) -> Unit,
) {
    val options: List<Pair<FeatureStatus?, String>> = listOf(
        null to "Populair",
        FeatureStatus.CONSIDERING to "In overweging",
        FeatureStatus.PLANNED to "Gepland",
        FeatureStatus.DONE to "Klaar",
        FeatureStatus.DECLINED to "Afgewezen",
    )
    LazyRow(
        contentPadding = androidx.compose.foundation.layout.PaddingValues(horizontal = Dvh.s4),
        horizontalArrangement = Arrangement.spacedBy(Dvh.s2),
    ) {
        items(options) { (status, label) ->
            val isSelected = status == selected
            Text(
                text = label,
                style = MaterialTheme.typography.labelLarge,
                color = if (isSelected) Color.White else Brand.Ink2,
                modifier = Modifier
                    .clip(CircleShape)
                    .background(if (isSelected) Brand.Moss else Brand.MossSoft)
                    .clickable { onSelect(status) }
                    .padding(horizontal = Dvh.s3, vertical = Dvh.s2),
            )
        }
    }
}

@Composable
private fun WishCard(
    request: FeatureRequest,
    voting: Boolean,
    onUpvote: () -> Unit,
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(Dvh.rLg))
            .background(Brand.Cream)
            .padding(Dvh.s4),
        horizontalArrangement = Arrangement.spacedBy(Dvh.s3),
    ) {
        Column(modifier = Modifier.weight(1f)) {
            Text(
                request.title,
                style = MaterialTheme.typography.titleMedium,
                color = Brand.Ink,
            )
            if (!request.body.isNullOrBlank()) {
                Text(
                    request.body!!,
                    style = MaterialTheme.typography.bodyMedium,
                    color = Brand.Ink2,
                    maxLines = 2,
                    overflow = TextOverflow.Ellipsis,
                    modifier = Modifier.padding(top = Dvh.s1),
                )
            }
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(Dvh.s2),
                modifier = Modifier.padding(top = Dvh.s2),
            ) {
                StatusPill(request.status)
                if (!request.component.isNullOrBlank()) {
                    Text(
                        request.component!!,
                        style = MaterialTheme.typography.labelSmall,
                        color = Brand.MossDark,
                        modifier = Modifier
                            .clip(CircleShape)
                            .background(Brand.MossSoft)
                            .padding(horizontal = Dvh.s2, vertical = 3.dp),
                    )
                }
            }
        }

        UpvoteButton(
            count = request.upvoteCount,
            active = request.viewerHasVoted,
            voting = voting,
            onClick = onUpvote,
        )
    }
}

@Composable
private fun UpvoteButton(
    count: Int,
    active: Boolean,
    voting: Boolean,
    onClick: () -> Unit,
) {
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        modifier = Modifier
            .clip(RoundedCornerShape(Dvh.rMd))
            .background(if (active) Brand.MossSoft else Color.Transparent)
            .border(
                width = 1.dp,
                color = if (active) Brand.Moss else Brand.Ink2.copy(alpha = 0.25f),
                shape = RoundedCornerShape(Dvh.rMd),
            )
            .clickable(enabled = !voting) { onClick() }
            .padding(horizontal = Dvh.s3, vertical = Dvh.s2),
    ) {
        if (voting) {
            CircularProgressIndicator(
                color = Brand.Moss,
                strokeWidth = 2.dp,
                modifier = Modifier.size(20.dp),
            )
        } else {
            Icon(
                Icons.Filled.KeyboardArrowUp,
                contentDescription = "Upvote",
                tint = if (active) Brand.Moss else Brand.Ink2,
                modifier = Modifier.size(22.dp),
            )
        }
        Text(
            "$count",
            style = MaterialTheme.typography.labelLarge,
            color = if (active) Brand.MossDark else Brand.Ink2,
        )
    }
}

@Composable
private fun StatusPill(status: FeatureStatus) {
    val (label, color) = when (status) {
        FeatureStatus.CONSIDERING -> "In overweging" to Brand.Terra
        FeatureStatus.PLANNED -> "Gepland" to Brand.Moss
        FeatureStatus.DONE -> "Klaar" to Brand.MossDark
        FeatureStatus.DECLINED -> "Afgewezen" to Brand.Rust
    }
    Text(
        text = label,
        style = MaterialTheme.typography.labelSmall,
        color = color,
        modifier = Modifier
            .clip(CircleShape)
            .background(color.copy(alpha = 0.12f))
            .padding(horizontal = Dvh.s2, vertical = 3.dp),
    )
}

@Composable
private fun NewWishSheet(
    onSubmit: (title: String, body: String?, component: String?, onResult: (String?) -> Unit) -> Unit,
) {
    var title by remember { mutableStateOf("") }
    var body by remember { mutableStateOf("") }
    var component by remember { mutableStateOf<String?>(null) }
    var submitting by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }

    val components = listOf("Kaart", "Inzenden", "Profiel", "Anders")
    val canSubmit = title.trim().length >= 3 && !submitting

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = Dvh.s4)
            .padding(bottom = Dvh.s6),
        verticalArrangement = Arrangement.spacedBy(Dvh.s3),
    ) {
        Text(
            "Nieuwe wens",
            style = MaterialTheme.typography.titleLarge,
            color = Brand.Ink,
        )

        val fieldColors = TextFieldDefaults.colors(
            focusedContainerColor = Brand.Sand,
            unfocusedContainerColor = Brand.Sand,
            focusedIndicatorColor = Brand.Moss,
            unfocusedIndicatorColor = Brand.Ink2.copy(alpha = 0.25f),
            cursorColor = Brand.Moss,
        )

        OutlinedTextField(
            value = title,
            onValueChange = { if (it.length <= 120) title = it },
            label = { Text("Titel") },
            singleLine = true,
            keyboardOptions = KeyboardOptions(imeAction = ImeAction.Next),
            modifier = Modifier.fillMaxWidth(),
            colors = fieldColors,
        )

        OutlinedTextField(
            value = body,
            onValueChange = { if (it.length <= 1000) body = it },
            label = { Text("Omschrijving (optioneel)") },
            minLines = 3,
            modifier = Modifier.fillMaxWidth(),
            colors = fieldColors,
        )

        Text("Onderdeel", style = MaterialTheme.typography.labelLarge, color = Brand.Ink2)
        LazyRow(horizontalArrangement = Arrangement.spacedBy(Dvh.s2)) {
            items(components) { c ->
                val isSelected = component == c
                Text(
                    text = c,
                    style = MaterialTheme.typography.labelLarge,
                    color = if (isSelected) Color.White else Brand.Ink2,
                    modifier = Modifier
                        .clip(CircleShape)
                        .background(if (isSelected) Brand.Moss else Brand.MossSoft)
                        .clickable { component = if (isSelected) null else c }
                        .padding(horizontal = Dvh.s3, vertical = Dvh.s2),
                )
            }
        }

        if (error != null) {
            Text(error!!, color = Brand.Rust, style = MaterialTheme.typography.bodyMedium)
        }

        Button(
            onClick = {
                submitting = true
                error = null
                onSubmit(title, body, component) { err ->
                    submitting = false
                    error = err
                }
            },
            enabled = canSubmit,
            colors = ButtonDefaults.buttonColors(containerColor = Brand.Moss),
            modifier = Modifier
                .fillMaxWidth()
                .height(Dvh.controlHeight),
        ) {
            if (submitting) {
                CircularProgressIndicator(
                    color = Color.White,
                    strokeWidth = 2.dp,
                    modifier = Modifier.size(20.dp),
                )
                Spacer(Modifier.width(Dvh.s2))
            }
            Text("Wens indienen")
        }
    }
}

@Composable
private fun CenterBox(content: @Composable () -> Unit) {
    Box(modifier = Modifier.fillMaxSize(), contentAlignment = Alignment.Center) { content() }
}
